File: binary_search/binary_search.go
package binary_search

func BinarySearch(haystack []int, needle int) int {
  lo := 0
  hi := len(haystack)

  for lo < hi {
    mi := lo + (hi - lo) / 2

    switch {
    case haystack[mi] == needle:
        return mi
    case haystack[mi] < needle:
        lo = mi + 1
    case haystack[mi] > needle:
        hi = mi
    }
  }

  return -1
}

File: src/binary_search.rs
use std::cmp::Ordering;

pub fn binary_search(haystack: &[i32], needle: i32) -> Option<usize> {
    let mut lo: usize = 0;
    let mut hi: usize = haystack.len();

    while lo < hi {
        let m: usize = lo + (hi - lo) / 2;

        match haystack[m].cmp(&needle) {
            Ordering::Equal => return Some(m),
            Ordering::Greater => lo = m + 1,
            Ordering::Less => hi = m,
        }
    }

    return None;
}
